### DISCUSSION TITLE
Bech32 encoding in age

### DISCUSSION THREAD

**Discussion by Tronic** [FiloSottile/age#352] [Q&A] [answered]
There seems to be something odd with the age key encryption. I have tried several implementations (Python packages bech32 and bitcoinlib) but neither seem to be able to to decode the addresses given in Age specs, nor to recreate the hash given for 32 times 0x42 (bitcoinlib actually plain refuses to convert this and some other specific sequences, although it generally accepts 32 random bytes).

Is the spec somehow customized for age, and where should I look for a usable implementation of it?

Thanks.

**Comment by FiloSottile** [comment]
Oh, I suspect those are not real examples. Try generating examples with one of the existing implementations, it should be standard Bech32. Sorry about that, a better spec is long overdue, but coming!

**Reply by Tronic** [reply]
The generated ones don't work either. I believe that any 32 byte random data should be okay, so it would seem that bitcoinlib is buggy because while it correctly recognizes the formats, it fails converting the bits (it would seem that it gets values out of range and then bails out while doing the 5-bit-8-bit conversions). The other library does not support 8 bit buffers but instead only uses lists of 5 bit numbers. How exactly these should be converted into binary is hard to find out when there are no sample keys that could be tested against (I could not match that example in Age spec no matter which endianness etc. I tried).

**Comment by Tronic** [comment, marked as answer]
Not a perfect solution but something that can be worked with, https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py

Need to use the bech32_encode/decode functions together with the convertbits function, and manually remove the 33rd byte (value 0) that it outputs. Also have to work with the secret key always in lower case or otherwise wrong checksum is created (but this is mentioned in Age spec).

**Reply by Tronic** [reply]
Setting padding False avoids the manual removal of bytes but I ended up rewriting much of it. One trick to know is that since the total number of bits in 5 bit array does not match exactly the 8 bit buffer, there must be unused zero bits on the lowest bits of the final element of the five bit array. In the case of 32 bytes, you have 4 bits unused bit and only the highest bit is used, i.e. the last number can be either 0 or 16 but nothing else. The numbers need left *and* right shifting to get it properly aligned to 8 bits.
